From 4559fc6fd570cde3ca1d6033fe63adf211ea2885 Mon Sep 17 00:00:00 2001
From: brokestar233 <3765589194@qq.com>
Date: Tue, 25 Jun 2024 07:18:27 +0000
Subject: [PATCH] ANDROID: gki_defconfig: Enable LLVM Polly optimizer for GKI
 builds

---
 Makefile     | 4 ++++
 arch/Kconfig | 8 ++++++++
 3 files changed, 12 insertions(+)

diff --git a/Makefile b/Makefile
index 4b4f51e1858a..1d1872b7a9e1 100644
--- a/Makefile
+++ b/Makefile
@@ -849,6 +849,10 @@ endif
 KBUILD_RUSTFLAGS += -Cdebug-assertions=$(if $(CONFIG_RUST_DEBUG_ASSERTIONS),y,n)
 KBUILD_RUSTFLAGS += -Coverflow-checks=$(if $(CONFIG_RUST_OVERFLOW_CHECKS),y,n)

+ifdef CONFIG_LLVM_POLLY
+KBUILD_CFLAGS	+= -mllvm -polly
+endif
+
 # Tell gcc to never replace conditional load with a non-conditional one
 ifdef CONFIG_CC_IS_GCC
 # gcc-10 renamed --param=allow-store-data-races=0 to
diff --git a/arch/Kconfig b/arch/Kconfig
index 5b2bd4f346b9..a9d821262d05 100644
--- a/arch/Kconfig
+++ b/arch/Kconfig
@@ -851,27 +851,35 @@
 	  If unsure, say N.

 config LLVM_POLLY
-	bool
-	prompt "Enable LLVM's polyhedral loop optimizer (Polly)"
+	bool "Enable LLVM's polyhedral loop optimizer (Polly)"
 	help
 	  This option enables LLVM's polyhedral loop optimizer known as Polly.
 	  Polly is able to optimize various loops throughout the kernel for
 	  maximum cache locality. This requires an LLVM toolchain explicitly
 	  compiled with Polly support.

 config HAVE_ARCH_WITHIN_STACK_FRAMES
-	bool
 	bool "Does the architecture support walking the stack on this function?"
 	help
 	  An architecture should select this if it can walk the kernel stack
-- 
2.43.0